contracts/SwapEthWrapper.sol

```
@@ -0,0 +1,298 @@
+     * represents WETH9 in the pool, the caller must set msg.value equal to dx.
+     * If the user is swapping to WETH9 in the pool, the user will receive ETH instead.
+     * @param tokenIndexFrom the token the user wants to swap from
+     * @param tokenIndexTo the token the user wants to swap to
+     * @param dx the amount of tokens the user wants to swap from
+     * @param minDy the min amount the user would like to receive, or revert.
+     * @param deadline latest timestamp to accept this transaction
+     */
+    function swap(
+        uint8 tokenIndexFrom,
+        uint8 tokenIndexTo,
+        uint256 dx,
+        uint256 minDy,
+        uint256 deadline
+    ) external payable returns (uint256) {
+        // Transfer tokens from msg.sender to this contract
+        if (tokenIndexFrom != wethIndex) {
+            IERC20(pooledTokens[tokenIndexFrom]).safeTransferFrom(
+                msg.sender,
+                address(this),
+                dx
+            );
+        } else {
+            IWETH9(weth).deposit{value: msg.value}();
+            require(msg.value == dx, "INCORRECT_MSG_VALUE");
+        }
+        // Execute swap
+        uint256 dy =
+            swapInstance.swap(
+                tokenIndexFrom,
+                tokenIndexTo,
+                dx,
+                minDy,
+                deadline
+            );
+        // Transfer the swapped tokens to msg.sender
+        if (tokenIndexTo != wethIndex) {
+            IERC20(pooledTokens[tokenIndexTo]).safeTransfer(msg.sender, dy);
+        } else {
+            IWETH9(weth).withdraw(dy);
+            // slither-disable-next-line arbitrary-send
+            (bool success, ) = msg.sender.call{value: dy}("");
+            require(success, "ETH_TRANSFER_FAILED");
+        }
+        return dy;
+    }
+
+    /**
+     * @notice Rescues any of the ETH, the pooled tokens, or the LPToken that may be stuck
+     * in this contract.
+     */
+    function rescue() external {
```

Anyone can call this?

---

Yes. I don't expect this to happen unless the transfers fail silently.

Though it would be a better practice to make it only callable by the multisig. Will change this in the next commit.